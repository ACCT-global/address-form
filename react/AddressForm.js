import React, { Component } from 'react'
import PropTypes from 'prop-types'
import AddressShapeWithValidation from './propTypes/AddressShapeWithValidation'
import InputFieldContainer from './InputFieldContainer'
import DefaultInput from './inputs/DefaultInput'
import {
  filterPostalCodeFields,
  filterAutoCompletedFields,
  isDefiningPostalCodeField,
} from './selectors/fields'
import SelectPostalCode from './postalCodeFrom/SelectPostalCode'
import { injectRules } from './addressRulesContext'
import { compose } from 'recompose'
import { injectAddressContext } from './addressContainerContext'
import GeolocationNumberForm from './GeolocationNumberForm'
import GoogleMapsContainer from './geolocation/GoogleMapsContainer'
import { injectIntl, intlShape } from 'react-intl'

class AddressForm extends Component {
  bindOnChange = () => {
    const { address, onChangeAddress, intl } = this.props
    return () => {
      onChangeAddress({
        'number': {
          ...address['number'],
          value: !address['number'].disabled
            ? intl.formatMessage({
              id: 'address-form.field.notApplicable',
            }) : '',
          disabled: !address['number'].disabled,
        },
      })
    }
  }

  render() {
    const {
      address,
      disableNumberInput,
      rules,
      onChangeAddress,
      Input,
      omitPostalCodeFields,
      omitAutoCompletedFields,
      geolocation,
      googleMapsKey,
      isNumberInputEnabled,
      onNumberInputFocus,
    } = this.props
    let fields = omitPostalCodeFields
      ? filterPostalCodeFields(rules)
      : rules.fields

    fields = omitAutoCompletedFields
      ? filterAutoCompletedFields({ fields }, address)
      : fields

    const fieldCondition = fields.length > 3
    const geolocationCondition = geolocation && fieldCondition

    if (!fieldCondition) {
      disableNumberInput()
    }

    return (
      <div className="flex items-center">
        {fields.map(
          (field, index) =>
            isDefiningPostalCodeField(field.name, rules) ? (
              <SelectPostalCode
                Input={Input}
                rules={rules}
                address={address}
                onChangeAddress={onChangeAddress}
              />
            ) : index === 0 && geolocationCondition ? (
              <GoogleMapsContainer apiKey={googleMapsKey}>
                {({ googleMaps }) => {
                  return (<GeolocationNumberForm
                    key={index}
                    onCheckedBox={this.bindOnChange()}
                    address={address}
                    rules={rules}
                    field={field}
                    onChangeAddress={onChangeAddress}
                    Input={Input}
                    omitPostalCodeFields={omitPostalCodeFields}
                    omitAutoCompletedFields={omitAutoCompletedFields}
                    geolocation={geolocation}
                    googleMaps={googleMaps}
                    isNumberInputEnabled={isNumberInputEnabled}
                    onNumberInputFocus={onNumberInputFocus}
                    disableNumberInput={disableNumberInput}

                  />)
                }}
              </GoogleMapsContainer>
            ) : (
              <div key={index}>
                <InputFieldContainer
                  key={field.name}
                  Input={Input}
                  field={field}
                  address={address}
                  rules={rules}
                  onChangeAddress={onChangeAddress}
                />
              </div>
            ),
        )}
      </div>
    )
  }
}

AddressForm.defaultProps = {
  omitPostalCodeFields: true,
  omitAutoCompletedFields: true,
  Input: DefaultInput,
  isNumberInputEnabled: false,
}

AddressForm.propTypes = {
  Input: PropTypes.func,
  address: AddressShapeWithValidation,
  disableNumberInput: PropTypes.func,
  omitPostalCodeFields: PropTypes.bool,
  omitAutoCompletedFields: PropTypes.bool,
  rules: PropTypes.object.isRequired,
  onChangeAddress: PropTypes.func.isRequired,
  geolocation: PropTypes.bool,
  googleMapsKey: PropTypes.string,
  isNumberInputEnabled: PropTypes.bool,
  intl: intlShape,
  onNumberInputFocus: PropTypes.func,
}

const enhance = compose(
  injectAddressContext,
  injectRules,
  injectIntl,
)
export default enhance(AddressForm)
