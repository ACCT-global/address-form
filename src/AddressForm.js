import React, { Component } from 'react'
import PropTypes from 'prop-types'
import AddressShape from './propTypes/AddressShape'
import { POSTAL_CODE, ONE_LEVEL, TWO_LEVELS, THREE_LEVELS } from './constants'
import filter from 'lodash/filter'
import InputSelect from './addressInputs/InputSelect'
import InputText from './addressInputs/InputText'
import { hasOptions } from './rulesLens/fields'

class AddressForm extends Component {
  constructor(props) {
    super(props)

    this.state = {
      fields: props.omitPostalCodeFields
        ? filterFields(props.rules)
        : props.rules.fields,
    }
  }

  componentWillReceiveProps(nextProps) {
    if (
      nextProps.rules.postalCodeFrom !== this.props.rules.postalCodeFrom ||
      nextProps.omitPostalCodeFields !== this.props.omitPostalCodeFields
    ) {
      this.setState({
        fields: nextProps.omitPostalCodeFields
          ? filterFields(nextProps.rules)
          : nextProps.rules.fields,
      })
    }
  }

  render() {
    const { address, rules, onChangeAddress } = this.props
    const { fields } = this.state

    return (
      <div>
        {fields.map(field => (
          <div key={field.name}>
            <label>
              {field.label}
              {hasOptions(field)
                ? <InputSelect
                  field={field}
                  rules={rules}
                  address={address}
                  onChange={onChangeAddress}
                  />
                : <InputText
                  field={field}
                  address={address}
                  onChange={onChangeAddress}
                  />}
            </label>
          </div>
        ))}
      </div>
    )
  }
}

AddressForm.defaultProps = {
  omitPostalCodeFields: true,
}

AddressForm.propTypes = {
  address: PropTypes.shape(AddressShape),
  omitPostalCodeFields: PropTypes.bool,
  rules: PropTypes.object.isRequired,
  onChangeAddress: PropTypes.func.isRequired,
}

/**
 * Remove fields already filled in PostalCodeGetter
 * @param {Object} rules
 */
export function filterFields(rules) {
  switch (rules.postalCodeFrom) {
    case THREE_LEVELS:
      return filter(
        rules.fields,
        ({ name }) => rules.postalCodeLevels.indexOf(name) === -1
      )
    case TWO_LEVELS:
      return filter(
        rules.fields,
        ({ name }) => rules.postalCodeLevels.indexOf(name) === -1
      )
    case ONE_LEVEL:
      return filter(rules.fields, ({ name }) => rules.postalCodeLevel !== name)
    default:
    case POSTAL_CODE:
      return filter(rules.fields, ({ name }) => name !== 'postalCode')
  }
}

export default AddressForm
