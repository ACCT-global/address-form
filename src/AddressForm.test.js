import React from 'react'
import { shallow } from 'enzyme'
import AddressForm from './AddressForm'

it('renders without crashing', () => {
  const wrapper = shallow(
    <AddressForm
      country={'BRA'}
      fields={[
        {
          name: 'street',
          value: 'Praia de Botafogo',
          valid: true,
          required: true,
        },
      ]}
      onChangeAddress={jest.fn()}
    />
  )
  const div = <div />
  expect(wrapper.contains(div)).toEqual(true)
})
