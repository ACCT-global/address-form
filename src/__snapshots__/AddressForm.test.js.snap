// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressForm should omit fields that are used in to get the postal code 1`] = `
<div>
  <p
    className="input ship-street required text"
  >
    <label
      htmlFor="ship-street"
    />
    <input
      className="input-xlarge"
      disabled={false}
      id="ship-street"
      name="street"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={null}
      type="text"
      value=""
    />
  </p>
  <p
    className="input ship-number required text"
  >
    <label
      htmlFor="ship-number"
    />
    <input
      className="input-mini"
      disabled={false}
      id="ship-number"
      name="number"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={null}
      type="text"
      value=""
    />
  </p>
  <p
    className="input ship-complement text"
  >
    <label
      htmlFor="ship-complement"
    />
    <input
      className="input-large"
      disabled={false}
      id="ship-complement"
      name="complement"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={undefined}
      type="text"
      value=""
    />
  </p>
  <p
    className="input ship-neighborhood required text"
  >
    <label
      htmlFor="ship-neighborhood"
    />
    <input
      className="input-large"
      disabled={false}
      id="ship-neighborhood"
      name="neighborhood"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={null}
      type="text"
      value=""
    />
  </p>
  <p
    className="input ship-city required text"
  >
    <label
      htmlFor="ship-city"
    />
    <input
      className="input-large"
      disabled={false}
      id="ship-city"
      name="city"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={null}
      type="text"
      value=""
    />
  </p>
  <p
    className="input ship-state required text"
  >
    <label
      htmlFor="ship-state"
    />
    <select
      className="input-mini"
      disabled={false}
      id="ship-state"
      name="state"
      onBlur={[Function]}
      onChange={[Function]}
      value=""
    >
      <option
        disabled={undefined}
        value=""
      >
        UF
      </option>
      <option
        value="AC"
      >
        Acre
      </option>
      <option
        value="AL"
      >
        Alagoas
      </option>
      <option
        value="AP"
      >
        Amap√°
      </option>
      <option
        value="RJ"
      >
        Rio de Janeiro
      </option>
    </select>
  </p>
  <p
    className="input ship-receiverName required text"
  >
    <label
      htmlFor="ship-receiverName"
    />
    <input
      className="input-large"
      disabled={false}
      id="ship-receiverName"
      name="receiverName"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={null}
      type="text"
      value=""
    />
  </p>
</div>
`;
